pkg/kubelet/cm/qos_container_manager_linux.go

```
@@ -0,0 +1,266 @@
+		if err != nil {
+			return err
+		}
+		usage := stats.MemoryStats.Usage
+
+		// Because there is no good way to determine of the original Update()
+		// on the memory resource was successful, we determine failure of the
+		// first attempt by checking if the usage is above the limit we attempt
+		// to set.  If it is, we assume the first attempt to set the limit failed
+		// and try again setting the limit to the usage.  Otherwise we leave
+		// the CgroupConfig as is.
+		if configs[qos].ResourceParameters.Memory != nil && usage > *configs[qos].ResourceParameters.Memory {
+			configs[qos].ResourceParameters.Memory = &usage
+		}
+	}
+
+	return nil
+}
+
+func (m *qosContainerManagerImpl) UpdateReserve() error {
+	if len(m.QOSReserveRequests) == 0 {
+		return nil
+	}
+
+	qosConfigs := map[v1.PodQOSClass]*CgroupConfig{
+		v1.PodQOSBurstable: &CgroupConfig{
+			Name:               CgroupName(m.qosContainersInfo.Burstable),
+			ResourceParameters: &ResourceConfig{},
+		},
+		v1.PodQOSBestEffort: &CgroupConfig{
+			Name:               CgroupName(m.qosContainersInfo.BestEffort),
+			ResourceParameters: &ResourceConfig{},
+		},
+	}
+
+	for resource, percentReserve := range m.QOSReserveRequests {
+		switch resource {
+		case v1.ResourceMemory:
+			m.setMemoryReserve(qosConfigs, percentReserve)
+		}
+	}
+	updateSuccess := true
+	for _, config := range qosConfigs {
+		err := m.cgroupManager.Update(config)
+		if err != nil {
+			updateSuccess = false
+		}
+	}
+	if updateSuccess {
+		glog.V(2).Infof("[Container Manager] QoS reserve requests set successfully")
+		return nil
+	}
+
+	// If the resource can adjust the ResourceConfig to increase likelihood of
+	// success, call the adjustment function here.  Otherwise, the Update() will
+	// be called again with the same values.
+	for resource, percentReserve := range m.QOSReserveRequests {
+		switch resource {
+		case v1.ResourceMemory:
+			m.retrySetMemoryReserve(qosConfigs, percentReserve)
```

also appear to ignore an error here as well, why?